library/std/src/sys/uefi/stdio.rs

```diff
@@ -0,0 +1,270 @@
+    ) -> io::Result<()> {
+        let r = unsafe {
+            let mut x: usize = 0;
+            (wait_for_event)(1, &mut (*con_in.as_ptr()).wait_for_key, &mut x)
+        };
+
+        if r.is_error() {
+            Err(io::Error::new(io::ErrorKind::Other, "Could not wait for event"))
+        } else {
+            Ok(())
+        }
+    }
+
+    // FIXME Improve Errors
+    fn read_key_stroke(con_in: NonNull<simple_text_input::Protocol>) -> io::Result<u16> {
+        let mut input_key = simple_text_input::InputKey::default();
+        let r = unsafe { ((*con_in.as_ptr()).read_key_stroke)(con_in.as_ptr(), &mut input_key) };
+
+        if r.is_error() || input_key.scan_code != 0 {
+            Err(io::Error::new(io::ErrorKind::InvalidInput, "Error in Reading Keystroke"))
+        } else {
+            Ok(input_key.unicode_char)
+        }
+    }
+
+    // FIXME Improve Errors
+    fn reset_weak(con_in: NonNull<simple_text_input::Protocol>) -> io::Result<()> {
+        let r = unsafe { ((*con_in.as_ptr()).reset)(con_in.as_ptr(), r_efi::efi::Boolean::TRUE) };
+
+        if r.is_error() {
+            Err(io::Error::new(io::ErrorKind::InvalidInput, "Device Error"))
+        } else {
+            Ok(())
+        }
+    }
+
+    // FIXME Improve Errors
+    fn write_character(
+        con_out: NonNull<simple_text_output::Protocol>,
+        character: ucs2::Ucs2Char,
+    ) -> io::Result<()> {
+        let mut buf: [u16; 2] = [character.into(), 0];
+        let r = unsafe { ((*con_out.as_ptr()).output_string)(con_out.as_ptr(), buf.as_mut_ptr()) };
+
+        if r.is_error() {
+            Err(io::Error::new(io::ErrorKind::InvalidInput, "Device Error"))
+        } else if character == ucs2::Ucs2Char::CR {
+            // Handle enter key
+            Self::write_character(con_out, ucs2::Ucs2Char::LF)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl io::Read for Stdin {
+    // Reads 1 UCS-2 character at a time and returns.
+    // FIXME: Implement buffered reading. Currently backspace and other characters are read as
+    // normal characters. Thus it might look like line-editing but it actually isn't
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if let Ok(current_exe) = crate::env::current_exe() {
+            if let Ok(v) = crate::env::var(format!("{}_stdin", current_exe.to_string_lossy())) {
+                if v.as_str() == "null" {
+                    return Ok(buf.len());
+                }
+            }
+        }
+
+        let global_system_table = uefi::env::get_system_table()
+            .ok_or(io::Error::new(io::ErrorKind::NotFound, "Global System Table"))?;
+        let con_in = get_con_in(global_system_table)?;
+        let con_out = get_con_out(global_system_table)?;
+        let wait_for_event = get_wait_for_event(global_system_table)?;
+
+        if buf.len() < 3 {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "Buffer too small"));
```

Comment:
This should not be an error, passing an empty buffer is valid.
```suggestion
            return Ok(0);
```